File: infrastructure/db-core/src/main/java/com/nmh/commerce/coupon/DurationPeriodPolicyEntity.java
package com.nmh.commerce.coupon;

import jakarta.persistence.DiscriminatorValue;
import jakarta.persistence.Entity;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.time.Duration;

@NoArgsConstructor(access = AccessLevel.PROTECTED)
@DiscriminatorValue("DURATION")
@Entity
public class DurationPeriodPolicyEntity extends ExpirationPeriodPolicyEntity {

    private Duration duration;

    private DurationPeriodPolicyEntity(CouponEntity coupon, Duration duration) {
        super(coupon);
        this.duration = duration;
    }

    public static DurationPeriodPolicyEntity from(DurationPeriodPolicy durationPeriodPolicy, CouponEntity coupon) {
        return new DurationPeriodPolicyEntity(coupon, durationPeriodPolicy.duration);
    }

    @Override
    public ExpirationPeriodPolicy toDomain() {
        return new DurationPeriodPolicy(duration);
    }
}

File: infrastructure/db-core/src/main/kotlin/com/nmh/commerce/coupon/DurationPeriodPolicyEntity.kt
package com.nmh.commerce.coupon

import jakarta.persistence.DiscriminatorValue
import jakarta.persistence.Entity
import java.time.Duration

@DiscriminatorValue("DURATION")
@Entity
class DurationPeriodPolicyEntity private constructor(
    coupon: CouponEntity,
    private val duration: Duration,
) : ExpirationPeriodPolicyEntity(coupon = coupon) {
    override fun toDomain(): ExpirationPeriodPolicy = DurationPeriodPolicy(duration)

    companion object {
        fun from(
            durationPeriodPolicy: DurationPeriodPolicy,
            coupon: CouponEntity,
        ): DurationPeriodPolicyEntity = DurationPeriodPolicyEntity(coupon, durationPeriodPolicy.duration)
    }
}
